### DISCUSSION TITLE
supabase onAuthStateChange not re-rendering useEffect

### DISCUSSION THREAD

**Discussion by jithinlal** [supabase/supabase#5563] [Questions]
[](https://stackoverflow.com/posts/71202204/timeline)I have the following custom hook useAuth, in which I am having a supabase.auth.onAuthStateChange event listener.

```
export const useAuth = () => {
  const [session, setSession] = useState(supabase.auth.session());

  const signOutUser = async () => {
    await supabase.auth.signOut();
    useAuthStore.setState({
      user: null,
    });
    setSession(null);
  };

   useEffect(() => {
    console.log({ session });
  }, [session]);

  supabase.auth.onAuthStateChange(async (event, session) => {
    if (session) {      
      const { error: fetchError, data } = await fetchUser(session);
      if (fetchError) {
        await signOutUser();
      } else {
        if (data.length === 0) {         
          setSession(session);            
        } else {            
          setSession(session);            
        }
      }      
    } else {
      setSession(null);
    }
  }); 

  return session;
};
```

Why is the useEffect not getting re-rendered even after the setSession got executed inside the onAuthStateChanged listeners?

**Comment by kangmingtay** [comment]
Have you tried removing the async keyword in front of the callback function passed into `onAuthStateChange`? Seems like `onAuthStateChange` accepts a callback which shouldn't be an async function (https://github.com/supabase/gotrue-js/blob/862996337c38876eaefead036871ea10df2e27bd/src/GoTrueClient.ts#L480). 

You can also check out our react example [here](https://github.com/supabase/gotrue-js/blob/master/example/react/src/App.js#L18) where we use `onAuthStateChange` to display the session information upon signing-in.

**Reply by jithinlalpr** [reply]
It's not working. I think the problem I am facing here is not related to sup abase authstatechange but more into react router dom. I am calling this custom hook on a react router Dom component

`
<Route path="/" element={<RequireAuth><Home/></RequireAuth>} />`

```
function RequireAuth({ children }) {
	const session = useAuth();
	console.log({ session });
	let location = useLocation();

	if (!session) {
		return <Navigate to='/login' state={{ from: location }} replace />;
	}

	return children;
}
```

**Reply by TheoOliveira** [reply]
were able to fix thast i was trying to use this snippet for react native and gives me warning over using hooks inside useEffect.
